### PR TITLE
Named tuple string key access

### PR DESCRIPTION
**Pull request by jhass**
Implements #2966

### CODE REVIEW

**Comment by asterite** [comment]
I'm almost 👍 on this, I need to check with @waj. I'd also need to add a rule to the compiler to simplify named tuple indexing with string literals, like it's done now with symbol literals (a very easy change)

File: spec/compiler/semantic/named_tuple_spec.cr [hunk 1]
```diff
@@ -48,6 +62,28 @@ describe "Semantic: named tuples" do
...
+      )) { char }
+  end
+
+  it "types nilable named tuple access (6)" do
+    assert_type(%(
+      t = {x: 1, y: 'a'}
+      t["foo"]?
+      )) { nil_type }
+  end
+
```
Comment (jhass) [inline spec/compiler/semantic/named_tuple_spec.cr]: Not sure if that's the right place for these specs, but couldn't find anything else.

File: src/compiler/crystal/semantic/call.cr [hunk 2]
```diff
@@ -467,7 +467,7 @@ class Crystal::Call
 
   def named_tuple_indexer_helper(args, arg_types, owner, instance_type, nilable)
     arg = args.first
-    if arg.is_a?(SymbolLiteral)
+    if arg.is_a?(SymbolLiteral) || arg.is_a?(StringLiteral)
```
Comment (asterite) [inline src/compiler/crystal/semantic/call.cr]: Not sure this will filter `arg` to the correct type.

In any case, really nice that you touched compiler code! 🎉

Comment (jhass) [inline src/compiler/crystal/semantic/call.cr]: Well, all we want from is `.value` and either responds to it.

Comment (asterite) [inline src/compiler/crystal/semantic/call.cr]: Yes, if `||` did type filtering :-)

Comment (jhass) [inline src/compiler/crystal/semantic/call.cr]: Oh right, I forgot here that it has to compile with 0.18 >.<

File: src/string.cr [hunk 3]
```diff
@@ -1742,6 +1742,18 @@ class String
...
 
+  # Returns a string where all chars in the given hash are replaced
+  # by the corresponding *tuple* values.
+  #
+  # ```
+  # "hello".gsub({'e' => 'a', 'l' => 'd'}) # => "haddo"
+  # ```
+  def gsub(tuple : NamedTuple)
+    gsub do |char|
+      tuple[char.to_s]? || char
```
Comment (bcardiff) [inline src/string.cr]: Should we have`Tuple#fetch` to allow proper non-fallbacking when the actual value is falsy?
Not particularly needed in this case, but It's been a week with `hash[key] || default` bugs on my side :-)

Maybe adding `Indexable#fetch`?
`Hash#fetch` currently uses `#find_entry`, so I don't expect to promote that method to the module.

Comment (jhass) [inline src/string.cr]: Should I change something here or do you consider it a separate change? (I think I do personally). Note that the same pattern is currently in use for the existing `Hash(Char, _)` overload.

Comment (bcardiff) [inline src/string.cr]: Yeah, let's do it in a separate commit and get this merged. Apologies for the delay @jhass

**Comment by asterite** [comment]
@jhass Excellent, thank you!
